Sources/when.swift

```diff
@@ -86,3 +86,88 @@ public func when<U, V>(pu: Promise<U>, _ pv: Promise<V>) -> Promise<(U, V)> {
 public func when<U, V, X>(pu: Promise<U>, _ pv: Promise<V>, _ px: Promise<X>) -> Promise<(U, V, X)> {
     return _when([pu.asVoid(), pv.asVoid(), px.asVoid()]).then(on: zalgo) { (pu.value!, pv.value!, px.value!) }
 }
+
+/**
+ Wait for all promises in a set to resolve.
+
+ For example:
+ 
+ func downloadFile(url: NSURL) -> Promise<NSData> {
+    // ...
+ }
+ 
+ let urls: [NSURL] = ...
+ let urlGenerator = urls.generate()
+ 
+ let generator = AnyGenerator<Promise<NSData>> {
+    guard url = urlGenerator.next() else {
+        return nil
+    }
+
+    return downloadFile(url)
+ }
+
+ when(generator, concurrently: 3)
+    .then { datum: [NSData] -> Void in
+        // ...
+    }
+
+ - Warning: If *any* of the provided promises reject, the returned promise is immediately rejected with that promise’s rejection. The error’s `userInfo` object is supplemented with `PMKFailingPromiseIndexKey`.
+ - Warning: In the event of rejection the other promises will continue to resolve and, as per any other promise, will either fulfill or reject. This is the right pattern for `getter` style asynchronous tasks, but often for `setter` tasks (eg. storing data on a server), you most likely will need to wait on all tasks and then act based on which have succeeded and which have failed, in such situations use `join`.
+ - Parameter promises: Generator of promises.
+ - Returns: A new promise that resolves when all the provided promises fulfill or one of the provided promises rejects.
+ - SeeAlso: `join()`
+ */
+
+public func when<T, PromiseGenerator: GeneratorType where PromiseGenerator.Element == Promise<T> >(promiseGenerator: PromiseGenerator, concurrently: Int = 1) -> Promise<[T]> {
+    var generator = promiseGenerator
+
+    return Promise { fulfill, reject in
+        var pendingPromises = 0
+        var values: [T?] = []
+        var rejected = false
+        var counter = 0
+
+        func extendArraySize(size: Int) {
+            if size > values.capacity {
```

Comment:
I just realized that probably I should not do it because standard library should extend capacity automatically.